ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -548,5 +552,41 @@ private void addDatasetVersion(BlobProperties blobProperties, RestRequest restRe
             RestServiceErrorCode.getRestServiceErrorCode(ex.getErrorCode()));
       }
     }
+
+    /**
+     * Support delete the dataset version out of retentionCount
+     */
+    private void deleteDatasetVersionOutOfRetentionCount() {
+      Dataset dataset = (Dataset) restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_DATASET);
+      try {
+        if (RestUtils.isDatasetVersionQueryEnabled(restRequest.getArgs())) {
+          String accountName = dataset.getAccountName();
+          String containerName = dataset.getContainerName();
+          String datasetName = dataset.getDatasetName();
+          List<DatasetVersionRecord> datasetVersionRecordList =
+              accountService.getAllValidVersionsOutOfRetentionCount(accountName, containerName, datasetName);
```

Comment:
That's a good concern. I think normally we should be good. Since we always delete once we successfully put a new version into the database. And we always leverage the same primary DB to do the get and delete. There's one case that when user failed to push the version and the version has not been deleted from DB after user successfully issue another new put, it will consider the failed put as a version and still delete the old one. But we never promise user that the version will be a valid version. Considering we already tell user that we are not guarantee to rollback the failed put version and it might need manual delete, so I think it should be good(this is how wormhole supported as well). But I think we could have some update in the future to add a new filed to the database when the put is done successfully, this might be better. Let me know if you have any concern about it.